gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java

```diff
@@ -162,6 +176,90 @@ public void testBackendConfiguration()
+
+        Request initiateRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthInitiatePath)
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "custom")
+                        .build();
+        Response initiateResponse = httpClient.newCall(initiateRequest).execute();
+        assertThat(isNullOrEmpty(initiateResponse.header("set-cookie"))).isFalse();
+
+        Request callbackRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthCallbackPath)
+                        .post(requestBody)
+                        .addHeader("Cookie", initiateResponse.header("set-cookie"))
+                        .build();
+        Response callbackResponse = httpClient.newCall(callbackRequest).execute();
+        assertThat(callbackResponse.body().string()).isEqualTo(oauthCallbackResponse);
+
+        Request logoutRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/custom/logout")
+                        .post(requestBody)
+                        .addHeader("Cookie", initiateResponse.header("set-cookie"))
+                        .build();
+        Response logoutResponse = httpClient.newCall(logoutRequest).execute();
+
+        assertThat(isNullOrEmpty(logoutResponse.header("set-cookie"))).isFalse();
+    }
+
+    @Test
+    public void testCookieSigning()
+            throws IOException
+    {
+        OkHttpClient httpClient = new OkHttpClient.Builder()
+                .connectTimeout(10, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(60, TimeUnit.SECONDS)
+                .build();
+        String oauthInitiateBody = "anything";
+        RequestBody requestBody =
+                RequestBody.create(
+                        MediaType.parse("application/json; charset=utf-8"), oauthInitiateBody);
+
+        Request initiateRequest =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + oauthInitiatePath)
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "custom")
+                        .build();
+        Response initiateResponse = httpClient.newCall(initiateRequest).execute();
+        String cookieHeader = initiateResponse.header("set-cookie");
+        assertThat(isNullOrEmpty(cookieHeader)).isFalse();
+        GatewayCookie cookie = GatewayCookie.fromCookie(new Cookie(cookieHeader.split("=")[0], cookieHeader.split("=")[1].split(";")[0]));
+        assertThat(isNullOrEmpty(cookie.getSignature())).isFalse();
+
+        // Tamper with values. This will cause the cookie to be ignored because its values will not match the signature,
+        // causing the request will be routed to the adhoc backend
```

Comment:
If cookie is invalid due to incorrect signature, we ignore the cookie and route it according to default routing logic?
Doesn't feel right about this. Any bug / mis-configuration could result in query being routed to random backend without error message. IMO, we should reject the request with invalid cookie.

Reply:
I think we should consider this in a more broader and consistent perspective, as how Gateway should do on invalid requests, soley as a load balancer routing with default conf on errors? or with some internal logic that checks and returns error without routing. 

Invalid requests can be from two cases, by Gateway and by Trino.
- The latter, of course, is bypassed without any decisions made by gateway.
- The former, as of now, seems to bypass and/or route to adhoc
  - e.g.) if user sends query request with wrong header, it is sent to adhoc (default)
  - e.g.) any NON HTTP Post request in white list path, including extra path by user, is forwarded. Other than that, it is considered an API for the gateway and could fail.
  
REST API is slighlty exception cases, but other than that i think sending to adhoc seems fine, keeping in mind that Gateway is a router.

Reply:
I agree with @oneonestar in this case, since the cookie will only have an invalid or missing signature due to misbehavior by the client. There are no cases where the client should need to interact with it, so I don't think we should try to handle this case gracefully.

Reply:
Sounds fair enough :)